### PR TITLE
Follow up to PR#13545: update testsuite.

### PR DESCRIPTION
**Pull request by quickfur**
Follow-up for #13545: Add tests for new reserved version identifier.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by quickfur** [comment]
Sorry for the last minute push, decided to include a doc update as well so that people are aware of exactly what `-profile=gc` tracks or doesn't track.

File: test/fail_compilation/reserved_version_switch.d [hunk 1]
```diff
@@ -315,5 +316,6 @@ Error: version identifier `all` is reserved and cannot be set
 Error: version identifier `none` is reserved and cannot be set
 Error: version identifier `D_PreConditions` is reserved and cannot be set
 Error: version identifier `D_PostConditions` is reserved and cannot be set
+Error: version identifier `D_ProfileGC` is reserved and cannot be set
```
Comment (ljmf00) [inline test/fail_compilation/reserved_version_switch.d:319]: The CI fails here. Did you rebase? If yes, the last change seems to be broken.

**Comment by maxhaton** [comment]
Is the plan with this to annotate literally all allocations in phobos? What is the version flag for?

**Comment by kinke** [comment]
> Is the plan with this to annotate literally all allocations in phobos?

Yeah, what's the plan here? Even if all templates in Phobos would use that version for calling the tracing allocators, that's still no guarantee that all instantiations use it (because Phobos itself, compiled without `-profile=gc`, instantiates some of its templates, and the compiler then might not emit such template instances in user-code compiled with `-profile=gc`).

Rebuilding druntime and Phobos with `-profile=gc` and linking that would make things work presumably, incl. non-templated code in those libs. Everything else seems like a hack.

**Comment by quickfur** [comment]
> Is the plan with this to annotate literally all allocations in phobos? What is the version flag for?

Only those that call the GC API directly, which so far, AFAIK, is only `arrayAllocImpl` in std.array. Anything allocated through language constructs don't need to be registered, for example, because the compiler already does that for us. Anything allocated with C malloc doesn't count since it's not GC-related.

**Comment by maxhaton** [comment]
And it's definitely impossible to just make the offending functions use the
GC via things the compiler can transform? That function is a smell IMO,
should phobos really be calling the GC implementation like that?

File: test/fail_compilation/reserved_version_switch.d (hunk 1 above)
Comment (quickfur) [reply test/fail_compilation/reserved_version_switch.d:320]: Oops, missed one flag in the test case. Should work now.

**Comment by quickfur** [comment]
> And it's definitely impossible to just make the offending functions use the GC via things the compiler can transform? That function is a smell IMO, should phobos really be calling the GC implementation like that?

I didn't really want to touch the implementation details of `arrayAllocImpl`, since that gets into the whole debate of whether Phobos should be invoking internal GC APIs like that. I didn't look into why it was done this way, and I'm not sure I want to, since I'm sure there's *some* reason for it, justifiable or not, but it'd just distract from getting the user complaint addressed. If you guys think of a better way of doing this, by all means rewrite this code. I just didn't want the perfect to become the enemy of the good.

The one place we probably *need* the `D_ProfileGC` version is the API functions in `core.memory.GC`. Unless we hardcode API function names into the compiler and magically rewrite all calls to them. But that feels a lot dirtier than this approach, though, and introduces IMO too much magic into the compiler that may subtly break things down the road.

**Review by ljmf00 — APPROVED** [review]
LGTM

**Pull request merged**